staging/src/k8s.io/client-go/discovery/discovery_client.go

```diff
@@ -329,9 +333,18 @@ func (d *DiscoveryClient) ServerVersion() (*version.Info, error) {
 
 // OpenAPISchema fetches the open api schema using a rest client and parses the proto.
 func (d *DiscoveryClient) OpenAPISchema() (*openapi_v2.Document, error) {
-	data, err := d.restClient.Get().AbsPath("/swagger-2.0.0.pb-v1").Do().Raw()
+	data, err := d.restClient.Get().AbsPath("/openapi/v2").SetHeader("Accept", mimePb).Do().Raw()
 	if err != nil {
-		return nil, err
+		if errors.IsForbidden(err) || errors.IsNotFound(err) {
```

Comment:
looks like the apiserver in 1.9 returns the same answer from `/openapi/v2` as it does from `/`

that means a superuser (or request against the unsecured port) won't get an HTTP error, it'll get a json doc that isn't the openapi document. this code needs to detect that and we should fix that bug in the apiserver so that an unhandled path returns a 404, and doesn't fall back to `/`

Reply:
@liggitt the apiserver seems to return a 404 with the `/` json doc as body, which follows HTTPbis spec
https://github.com/kubernetes/kubernetes/blob/1d6e114ce7a090d80b87b6eb0ca88d4eca98de7f/staging/src/k8s.io/apiserver/pkg/server/config.go#L517-L520

< HTTP/1.1 404 Not Found
< Content-Length: 524
< Content-Type: application/json
< Date: Fri, 23 Feb 2018 20:04:32 GMT
< 
{
  "paths": [
    "/apis",
    "/apis/",
    "/apis/apiextensions.k8s.io",
    "/apis/apiextensions.k8s.io/v1beta1",
    "/healthz",
    "/healthz/etcd",
    "/healthz/ping",
    "/healthz/poststarthook/generic-apiserver-start-informers",
    "/healthz/poststarthook/start-apiextensions-controllers",
    "/healthz/poststarthook/start-apiextensions-informers",
    "/metrics",
    "/swagger-2.0.0.json",
    "/swagger-2.0.0.pb-v1",
    "/swagger-2.0.0.pb-v1.gz",
    "/swagger.json",
    "/swaggerapi",
    "/version"
  ]